### PR TITLE
Add load tests to run against the liquidator

### PR DESCRIPTION
**Pull request by dnut**
This adds a test that creates tokens, margin pools, swap pools, and users. The users deposit and borrow, then prices for half the assets are lowered gradually so that some of those users become unhealthy. This is intended to run against a localnet alongside a liquidator to see how well the liquidator performs. Number of tokens, users, and speed of price change are parameterized.

### TransactionBuilder
One part of this PR is adding the idea of a TransactionBuilder.
https://github.com/jet-lab/jet-v2/pull/211/files#diff-27d2d33468719783e1e8c44495e76d2e7fe520d16736bde6cc134383b762e00c

Previously, functions representing user actions would either return a `Transaction`, which cannot be modified and remain valid, or it would actually send a transaction to an RPC node. This requires an RPC client which is an unnecessary dependency imposed on instruction building logic, plus it doesn't allow the caller to combine the instructions in this transaction with other instructions in the same transaction, and it makes it a time sensitive operation (you need to send the tx shortly after selecting the blockhash).

With my change, some of these functions now return a `TransactionBuilder` struct, which includes a vec of instructions and a vec of signers. This communicates to callers that it is a bundle of instructions that are expected to be executed within the same transaction, and it carries all the necessary signers so the top level transaction executor doesn't need knowledge about the details of each instruction contained. `TransactionBuilder`s can be easily concatenated.

This allows a client to create their own transactions at any time by combining groups of instructions as they please, without requiring any additional information outside the TransactionBuilder, and compiling the transaction to be sent at any time.

<details><summary>Further Discussion - Alternative Approaches</summary>

- Simply return a vec of `Instruction`
  - Ambiguous when instructions need to be executed together or can be separated
  - Who are the signers for these instructions?
- Instead of `Vec<Instruction>` and a vec of signers, use `Vec<InstructionBuilder>` structs that contain a single `Instruction` with a vec of signers 
  - This is nice because it directly links the signer requirements along with the source of those requirements - the instruction. The transaction-layer requirements actually don't have any information about signers (other than payer). But with my approach there is a discrepancy: you need to include special logic about signers at the transaction layer, even though the information comes from the instruction layer. A straightforward design would allow the instruction layer to determine the signers required for that instruction.
  - The InstructionBuilder approach is a challenge, however, because most code is written with the assumption that signers are included at the transaction level. Existing instruction constructors often don't accept a signer parameter, only pubkeys. So I think the current TransactionBuilder approach is just a bit easier to adapt from existing patterns. Though the InstructionBuilder approach may be a better design in the long run.
  - A solution? The current TransactionBuilder approach actually allows the flexibility of taking either approach. You can write a signer-aware instruction constructor that simply treats the TransactionBuilder as an InstructionBuilder. It wraps its single instruction in a TransactionBuilder including all the necessary signers, and it can easily be joined with other TransactionBuilders on the caller's path to creating the actual transaction.

</details>

### CODE REVIEW

File: libraries/rust/src/solana/transaction.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+        Self { ..self }
+    }
+}
+
+const MAX_TX_SIZE: usize = 1232;
+
+/// Combines all the instructions within each of the TransactionBuilders into the smallest
+///  possible number of TransactionBuilders that don't violate the rules:
+/// - instructions that were grouped must in a TransactionBuilder must end up in the same TransactionBuilder
```
[the remark this answers is not in the document]
Comment (dnut) [reply libraries/rust/src/solana/transaction.rs:76]: instructions that were already grouped in a TransactionBuilder must end up in the same TransactionBuilder

File: tests/hosted/src/swap.rs [hunk 2]
```diff
@@ -41,6 +48,47 @@ related_programs! {
     ]}
 }
 
+pub type SwapRegistry = HashMap<Pubkey, HashMap<Pubkey, SwapPool>>;
```
[the remark this answers is not in the document]
Comment (dnut) [reply tests/hosted/src/swap.rs:51]: I expect your way to be faster, and simpler to insert values. My way makes it simpler to iterate once per token, or iterate over all pools for a given token.

File: tests/hosted/src/setup_helper.rs [hunk 3]
```diff
@@ -26,57 +33,98 @@ const DEFAULT_POOL_CONFIG: MarginPoolConfig = MarginPoolConfig {
...
+        token,
+        collateral_weight,
+        max_leverage: leverage_max,
+        token_kind: TokenKind::Collateral,
+        config: DEFAULT_POOL_CONFIG,
+        oracle: token_oracle,
+    };
+    let price = TokenPrice {
+        exponent: -8,
+        price: (price * 100_000_000.0) as i64,
```
[the remark this answers is not in the document]
Comment (dnut) [reply tests/hosted/src/setup_helper.rs:65]: Yeah floats may not always be appropriate, but when their quirks are acceptable I find them to be more user friendly, so I tried to abstract this manual exponent stuff into a small corner of the code.

**Comment by jetjake** [comment]
Was able to run tests on my local after doing this https://github.com/jet-lab/jet-v2/compare/load-test...load-test-testing?expand=1 status default lifetimes  with a longer sleep after the solana test validator

**Review by jetjake — APPROVED** [review]
Good stuff @dnut.  The load test is also a great example for extensions of that same idea where we can create all sorts of test cases and user types that do things to respond to the price movements.  But that's for another time.

**Pull request merged**